src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,407 @@
+            this.useGraph = true;
+        } else {
+            this.useGraph = false;
+        }
+
+        if (backEndType.equals(BackEndType.LOGIQL)) {
+            // Configure solving strategy.
+            this.solveInParallel = false;
+        } else if (solveInParallel == null || solveInParallel.equals(Constants.TRUE)) {
+            this.solveInParallel = true;
+        } else {
+            this.solveInParallel = false;
+        }
+
+        if (collectStatistic == null || collectStatistic.equals(Constants.FALSE)) {
+            // Configure statistic collection.
+            this.collectStatistic = false;
+        } else if (collectStatistic.equals(Constants.TRUE)) {
+            this.collectStatistic = true;
+        }
+
+        // Sanitize the configuration if it needs.
+        sanitizeConfiguration();
+        System.out.println("Configuration: \nback end type: " + this.backEndType.simpleName + "; \nuseGraph: "
+                + this.useGraph + "; \nsolveInParallel: " + this.solveInParallel + ".");
+    }
+
+    protected void configureLattice(QualifierHierarchy qualHierarchy) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        lattice = latticeBuilder.buildLattice(qualHierarchy);
+    }
+
+    protected TwoQualifiersLattice createTwoQualifierLattice(AnnotationMirror top, AnnotationMirror bottom) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        return latticeBuilder.buildTwoTypeLattice(top, bottom);
+    }
+
+    /**
+     * This method creates a ConstraintSerializer, which can deliver the
+     * constraint/slot to the real serializer. If a costume serialization logic
+     * is needed, user can have a subclass of ConstraintSerializer, override the
+     * serialize method with the costume logic, and then override this method
+     * with returning the instance of new subclass.
+     * 
+     * @param backEndType
+     * @param lattice
+     * @return A deliverer Serializer.
+     */
+    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
+        return new ConstraintSerializer<>(backEndType, lattice);
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    protected BackEnd<?, ?> createBackEnd(BackEndType backEndType, Map<String, String> configuration,
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment,
+            Lattice lattice, Serializer<?, ?> defaultSerializer) {
+
+        BackEnd<?, ?> backEnd = null;
+
+        try {
+            Class<?> backEndClass = Class.forName(backEndType.fullyQualifiedName + "BackEnd");
```

Comment:
In my opinion, the first benefit would be clearer code and improve reusability of the serializer creation logic. As my previous example, in sub-class of `GeneralSolver`, with `BackEndType` provide the creation logic, we can have code like:

```java
protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
        switch(backEndType) {
        //customize the backend type we are interested in
        case MAX_SAT: return new customMAXSATSerialzer();
        ....
        // in default, fall-back to general defaulting logic
        default: return backEndType.createDefaultSerializer();
        }
    }
```

Although it is functionally equal to the code calling `super.createSearilizer()` if we kept creation logic in `GeneralSolver`, above code is still more self-explain than `super.createSerializer()` as we can ease read the code and understand it is doing "using the default serializer". 

The other reason might be having a better code structure and let new comers easier to understand code. For illustration, `BackEndType` would be the detail place for defining what a `BackEnd` and `Serializer` should be: E.g. 1. what information do we need to create a new BackEnd? (simpleName, BackEnd.class, Serializer.class) 2. what constructor shall we have in our new BackEnd? (loading logic provide this information).

Then, on the solver level, the code there would focus on higher-level logic like how are we expected to associate BackEnd with Serializer? (plug serializer into BackEnd). What is the error-handling mechanism on the solver framework?  (try-catch the initialization error thrown from `BackEndType.createBackEndOrSerializer()`.